packages/language/src/typesystem/type-builder.ts

```diff
@@ -0,0 +1,734 @@
+import { Error } from "../validation/messages/pli-codes";
+import {
+  DataType,
+  DataTypes,
+  DataTypesByAttributeKind,
+  AttributeKinds,
+  AttributeTypes,
+  TypeDescriptions,
+  AttributeKind,
+  ScaleMode,
+  Volatility,
+  AlignmentType,
+  Assignability,
+  StorageClass,
+  ScopeType,
+  NumberMode,
+  Base,
+  StorageConnection,
+  StringKind,
+  StringFormat,
+  Endianess,
+  Sign,
+  AttributeWitnesses,
+  FloatFormat,
+  BufferMode,
+  AccessMode,
+  FileUsage,
+  Alignment,
+  Scope,
+} from "./descriptions";
+
+function createEmptyAttributeWitnesses(): AttributeWitnesses {
+  const obj: Partial<AttributeWitnesses> = {};
+  for (const kind of AttributeKinds) {
+    obj[kind] = null;
+  }
+  return obj as AttributeWitnesses;
+}
+
+type BuiltType = {
+  type: TypeDescriptions.Any | undefined;
+  diagnostics: Diagnostic[];
+};
+
+export interface TypeBuilder {
+  addAttribute(attribute: ast.DeclarationAttribute): void;
+  build(): BuiltType;
+}
+
+export class DefaultTypeBuilder implements TypeBuilder {
+  private diagnostics: Diagnostic[] = [];
+  private possibleDataTypes = new Set<DataType>(DataTypes);
+  private attributeWitnesses: AttributeWitnesses =
+    createEmptyAttributeWitnesses();
+  constructor(private token: Token | null) {}
+  addAttribute(attribute: ast.DeclarationAttribute): void {
+    switch (attribute.kind) {
+      case ast.SyntaxKind.ComputationDataAttribute:
+        if (attribute.type) {
+          this.handleDefaultAttribute(attribute);
+        }
+        break;
+      case ast.SyntaxKind.DateAttribute:
+      case ast.SyntaxKind.DefinedAttribute:
+        break;
+      case ast.SyntaxKind.DimensionsDataAttribute:
+        break;
+      case ast.SyntaxKind.EntryAttribute:
+      case ast.SyntaxKind.EnvironmentAttribute:
+      case ast.SyntaxKind.GenericAttribute:
+      case ast.SyntaxKind.HandleAttribute:
+      case ast.SyntaxKind.IndForAttribute:
+      case ast.SyntaxKind.InitialAttribute:
+      case ast.SyntaxKind.LikeAttribute:
+      case ast.SyntaxKind.OrdinalTypeAttribute:
+        break;
+      case ast.SyntaxKind.PictureAttribute:
+        /**
+         * Picture wideness attributes
+         * @see https://www.ibm.com/docs/en/epfz/6.1.0?topic=attributes-picture-widepic
+         */
+        break;
+      case ast.SyntaxKind.ReturnsAttribute:
+      case ast.SyntaxKind.TypeAttribute:
+      case ast.SyntaxKind.ValueAttribute:
+      case ast.SyntaxKind.ValueListAttribute:
+      case ast.SyntaxKind.ValueListFromAttribute:
+      case ast.SyntaxKind.ValueRangeAttribute:
+        break;
+      default:
+        assertUnreachable(attribute);
+    }
+  }
+  handleDefaultAttribute(attribute: ast.ComputationDataAttribute) {
+    const token = attribute.typeToken!;
+    assertType<ast.DefaultAttribute>(attribute.type);
+    switch (attribute.type) {
+      /**
+       * Data type attributes
+       */
+      case "TASK":
+      case "FILE":
+      case "FORMAT":
+      case "AREA": {
```

Comment:
Suggestion: Before continuing on that, can we refactor this to use an enum? The strings are pretty error-prone and difficult to maintain (they are an artifact of migrating the project from Langium). Ideally, we would have a mapping from token type -> `DefaultAttribute` enum value in the parsing phase.